runtime/src/chpl-gpu.c

```diff
@@ -194,5 +248,90 @@ void chpl_gpu_launch_kernel_flat(const char* name, int grd_dim, int blk_dim,
   va_end(args);
 }
 
+bool chpl_gpu_has_context() {
+  CUcontext cuda_context = NULL;
+
+  CUresult ret = cuCtxGetCurrent(&cuda_context);
+
+  if (ret == CUDA_ERROR_NOT_INITIALIZED || ret == CUDA_ERROR_DEINITIALIZED) {
+    return false;
+  }
+  else {
+    return cuda_context != NULL;
+  }
+}
+
+void* chpl_gpu_mem_alloc(size_t size, chpl_mem_descInt_t description,
+                         int32_t lineno, int32_t filename) {
+  CHPL_GPU_LOG("chpl_gpu_mem_alloc called. Size:%d file:%d line:%d\n", size,
+               filename, lineno);
+
+  CUdeviceptr ptr;
+  CUDA_CALL(cuMemAllocManaged(&ptr, size, CU_MEM_ATTACH_GLOBAL));
+
+  CHPL_GPU_LOG("chpl_gpu_mem_alloc returning %p\n", (void*)ptr);
+
+  return (void*)ptr;
+
+}
+
+void* chpl_gpu_mem_calloc(size_t number, size_t size,
+                          chpl_mem_descInt_t description,
+                          int32_t lineno, int32_t filename) {
+  CHPL_GPU_LOG("chpl_gpu_mem_calloc called. Size:%d\n", size);
+
+  CUdeviceptr ptr;
+  CUDA_CALL(cuMemAllocManaged(&ptr, size, CU_MEM_ATTACH_GLOBAL));
+  CUDA_CALL(cuMemsetD8(ptr, 0, size));
+  return (void*)ptr;
+}
+
+void* chpl_gpu_mem_realloc(void* memAlloc, size_t size,
+                           chpl_mem_descInt_t description,
+                           int32_t lineno, int32_t filename) {
+  CHPL_GPU_LOG("chpl_gpu_mem_realloc called. Size:%d\n", size);
+
+  assert(chpl_gpu_is_device_ptr(memAlloc));
+
+  size_t cur_size = chpl_gpu_get_alloc_size(memAlloc);
+
+  if (size == cur_size) {
+    return memAlloc;
+  }
+
+  // TODO we could probably do something smarter, especially for the case where
+  // the new allocation size is smaller than the original allocation size.
+
+  CUdeviceptr ptr;
+  CUDA_CALL(cuMemAllocManaged(&ptr, size, CU_MEM_ATTACH_GLOBAL));
+  CUDA_CALL(cuMemcpyDtoD(ptr, (CUdeviceptr)memAlloc, cur_size));
+  CUDA_CALL(cuMemFree((CUdeviceptr)memAlloc));
+
+  return (void*)ptr;
+}
+
+void* chpl_gpu_mem_memalign(size_t boundary, size_t size,
+                            chpl_mem_descInt_t description,
+                            int32_t lineno, int32_t filename) {
+  CHPL_GPU_LOG("chpl_gpu_mem_memalign called. Size:%d\n", size);
+  chpl_internal_error("Not ready to allocate aligned memory on GPU, yet.");
+
+  // ENGIN: I don't know if it is possible to allocate memory with custom
+  // alignment on GPU. It looks like GPU's typically have a default alignment
+  // (512?) that cannot be changed. I don't think we'd need more than that
+  // today, and if we want, we can play some pointer games to return something
+  // with a smaller alignment here.
```

Comment:
in the comment, do you mean "with a larger alignment" ?

Reply:
Yes, indeed. Good catch.